synapse/models/infotech.py

```diff
@@ -1,7 +1,150 @@
+    '''
+    return bruteVersion(valu)[0]
+
+def bruteVersion(valu):
+    '''
+    Attempt to brute force a valu into a semantic version string and its components
+
+    Args:
+        valu: A string or integer to attempt to obtain a system normalized version
+        valu and subs for.
+
+    Returns:
+        int, dict: The system normalized version integer and a subs dictionary.
+    '''
+    if isinstance(valu, int):
+        return s_datamodel.tlib.getTypeNorm('it:semver', valu)
+
+    if isinstance(valu, str):
+        return bruteStr(valu)
+
+    else:
+        raise s_common.BadTypeValu(valu=valu,
+                                   mesg='Unable to brute force  a valu',
+                                   type=type(valu))
+
+def bruteStr(valu):
+    '''
+    Brute force the version out of a string.
+
+    Args:
+        valu (str): String to attempt to get version information for.
+
+    Notes:
+        This first attempts to parse strings using the it:semver normalization
+        before attempting to extract version parts out of the string.
+
+    Returns:
+        int, dict: The system normalized version integer and a subs dictionary.
+    '''
+    try:
+        valu, subs = s_datamodel.tlib.getTypeNorm('it:semver', valu)
+        return valu, subs
+    except s_common.BadTypeValu:
+        # Try doing version part extraction by noming through the string
+        subs = s_version.parseVersionParts(valu)
+        if subs is None:
+            raise s_common.BadTypeValu(valu=valu,
+                                       mesg='Unable to brute force version parts out of the string')
+        if subs:
+            valu = s_version.packVersion(subs.get('major'),
+                                         subs.get('minor', 0),
+                                         subs.get('patch', 0))
+            return valu, subs
 
 class ItMod(CoreModule):
 
+    def initCoreModule(self):
+        self.onFormNode('it:dev:str', self._onFormItDevStr)
+        self.onFormNode('it:prod:softver', self._onFormItSoftVer)
+        self.core.addTypeCast('it:version:brute', bruteVersionValu)
+
+    def _onFormItDevStr(self, form, valu, props, mesg):
+        props['it:dev:str:norm'] = valu.lower()
+
+    def _onFormItSoftVer(self, form, valu, props, mesg):
+        # Set the :software:name field
+        if 'it:prod:softver:software:name' not in props:
+            software = props.get('it:prod:softver:software')
+            node = self.core.getTufoByProp('it:prod:soft', software)
+            if node:
+                name = node[1].get('it:prod:soft:name')
+                props['it:prod:softver:software:name'] = name
+        # Normalize the version string
+        vers = props.get('it:prod:softver:vers')
+        props['it:prod:softver:vers:norm'] = self.core.getTypeNorm('str:lwr', vers)[0]
+        if 'it:prod:softver:semver' in props:
+            return
+        # This could through a BadTypeValu and end the callback
```

Comment:
should we catch the exception then? or is there nothing we can do anyway?

Reply:
fixed with a try/except